Added support for UNIX sockets in HashClient.

pope1ni opened the pull request:
`HashClient` didn't support using paths to UNIX sockets and passing them to the underlying client.
With these changes it should also be possible to use a mix of UNIX sockets and `(host, port)` 2-tuples.

pope1ni commented:
Yes, I thought about doing this. Would you also like a `DeprecationWarning` to be triggered if `port` is set?

jparise commented:
> Yes, I thought about doing this. Would you also like a `DeprecationWarning` to be triggered if `port` is set?

`port` would still be used for TCP-based connections, as it is today, so there isn't a deprecation case that we need to announce.

pope1ni commented:
Ok. I've restored the original signature to those methods and added an extra test.

jparise commented on pymemcache/test/test_client_hash.py:
```diff
@@ -354,7 +355,27 @@ class MyClient(Client):
+            ('127.0.0.1', 11211),
+        ]
+        client = HashClient(servers)
+        assert set(servers) == set(c.server for c in client.clients.values())
```
You can simplify the right-hand side using a set comprehension:

```suggestion
        assert set(servers) == {c.server for c in client.clients.values()}
```

pope1ni replied:
:man_facepalming:

pope1ni commented on pymemcache/client/hash.py:
```diff
@@ -101,30 +102,41 @@ def __init__(
+        # To maintain backward compatibility, if a port is provided, assume
+        # that server wasn't provided as a (host, port) tuple.
+        if port is not None:
+            server = (server, port)
```
Sure. I've also updated to use `six.text_type` as I note that we're still supporting Python 2.

pope1ni commented:
Ok. I've made those changes. While I was at it I added a small cleanup for `isinstance()` checks noticed a bug in support for custom client classes when `use_pooling=True` for `HashClient` which I've also fixed.

jparise commented on pymemcache/test/test_client_hash.py:
```diff
@@ -42,7 +43,7 @@ def test_setup_client_without_pooling(self):
         with mock.patch(client_class) as internal_client:
             client = HashClient([], timeout=999, key_prefix='foo_bar_baz')
-            client.add_server('127.0.0.1', '11211')
+            client.add_server(('127.0.0.1', '11211'))
```
It doesn't look like we have any tests that use the `.add_server(server, port=port)` signature now?

pope1ni replied:
https://github.com/pinterest/pymemcache/pull/295/files#diff-dfcf37102448920749549449cc837e8aR382-R399

jparise replied:
Perfect. Thanks for the link!

pope1ni commented:
Hi @jparise. Would it be possible to make a release so that this fix is available? Thanks.

jparise commented:
@pope1ni yes, I'll run through the release process this morning.

pope1ni commented:
@jparise Thank you!
